core/src/consensus/tendermint/mod.rs

```diff
@@ -734,6 +736,7 @@ impl ConsensusEngine<CodeChainMachine> for Tendermint {
                 return Err(EngineError::DoubleVote(sender))
             }
             ctrace!(ENGINE, "Handling a valid {:?} from {}.", message, sender);
+            let _guard = self.step_change_lock.lock();
```

Comment:
Is it okay just locking here?
It seems that there are many paths to call `to_step` method.

Reply:
The problem was that `to_next_height` was called twice in `handle_valid_message`, so height is increased twice. I fixed it by not calling `handle_valid_message` concurrently.
As you say, although we locked the `handle_valid_message`, `to_step` is called concurrently and maybe it causes some problem. I'll look into what happens if the `to_step` is called concurrently.